src/CountdownBarrier.php

```diff
@@ -0,0 +1,124 @@
+<?php
+
+namespace Amp\Sync;
+
+use Amp\Deferred;
+use Amp\Promise;
+use Error;
+
+/**
+ * A countdown barrier returns a promise that is resolved when it was signaled n times.
+ *
+ * **Example**
+ *
+ * ```php
+ * $countdownBarrier = new \Amp\CountdownBarrier(2);
+ * $countdownBarrier->signal();
+ * $countdownBarrier->signal(); // promise is now resolved
+ * ```
+ */
+final class CountdownBarrier
+{
+    /** @var int */
+    private $initialCount;
+    /** @var int */
+    private $currentCount;
+    /** @var Deferred */
+    private $deferred;
+
+    public function __construct(int $initialCount)
+    {
+        if ($initialCount < 1) {
+            throw new Error('Counter must be positive');
+        }
+
+        $this->initialCount = $initialCount;
+        $this->currentCount = $initialCount;
+        $this->deferred = new Deferred();
+    }
+
+    public function getCurrentCount(): int
+    {
+        return $this->currentCount;
+    }
+
+    public function getInitialCount(): int
+    {
+        return $this->initialCount;
+    }
+
+    public function signal(int $signalCount = 1): bool
```

Comment:
In my opinion this should return void. If you need to check that the signal resolved the promise, you can call `getCurrentCount` right after.